### ISSUE TITLE
javaCC inserts non-wanted statements

### ISSUE THREAD

**Issue by lindenb** [javacc/javacc#24]
cross posted at http://stackoverflow.com/questions/43469509/

I've got the following javacc grammar.

The rule UnaryExpr creates some anonymous classes implementing `java.util.function.Predicate<>`

```java

options {
static=false;
DEBUG_PARSER=false;
IGNORE_CASE=true;
JDK_VERSION="1.8";
}

(...)

private Predicate<SAMRecord> UnaryExpr(): { }
    {
    (
    <DUPLICATE> { return new Predicate<SAMRecord>() {
            @Override public boolean test(final SAMRecord rec) { return rec.getDuplicateReadFlag();}
        };}
    | <UNMAPPED> { return new Predicate<SAMRecord>() {
            @Override public  boolean test(final SAMRecord rec) { return rec.getReadUnmappedFlag();}
        };}
    | <FAILSVENDORQUALITY> { return new Predicate<SAMRecord>() {
            @Override public  boolean test(final SAMRecord rec) { return rec.getReadFailsVendorQualityCheckFlag();}
        };}
    | <PROPERPAIR> { return new Predicate<SAMRecord>() {
            @Override public boolean test(final SAMRecord rec) { return rec.getReadPairedFlag();}
        };}
    )
    }
```

when I'm generating the code using **javacc 7.0.2**, the generated java code contains some extra statements that break the code {if ("" != null). In the java file, instead or my original code:

`
@Override public boolean test(final SAMRecord rec) {return rec.getDuplicateReadFlag();}
`

I get :

`@Override public boolean test(final SAMRecord rec) { {if ("" != null) return rec.getDuplicateReadFlag();}}`

How can I fix this ? Thanks .

PS: I've also tried to use java 8 lambda, but the java parser doesn't handle the following code: `<UNMAPPED> { return rec->rec.getReadUnmappedFlag();}`

**Comment by new-javacc** [comment]
About the first issue - looks like you need to set the option:

LEGACY_EXCEPTION_HANDLING = false

to suppress the "" != null

As for the second issue - JavaCC currently does not support java 8 
syntax. But my general suggestion to users is to avoid doing complex 
stuff in the grammar and use a parser subclass or superclass to do these 
things and make the parser actions to be simple method calls.

Hope that helps.

Sreeni.

**Comment by lindenb** [comment]
@new-javacc thanks for the quick answer. Bu it seems that this option doesn't exist anymore :
```

Java Compiler Compiler Version 7.0.2 (Parser Generator)
(type "javacc" with no arguments for help)
Reading from file src/main/resources/javacc/com/github/lindenb/jvarkit/util/bio/samfilter/SamFilterParser.jj . . .
Warning: Line 6, Column 29: Bad option value "false" for "LEGACY_EXCEPTION_HANDLING".  Option setting will be ignored.

```

**Comment by new-javacc** [comment]
Hmm, interesting. Looks like it was added sloppily. It seems to accept 
it from the command line, not in the file. I will look into it.

In any case, the if ("" != null) check should be true always so I don't 
see an issue with it. Why do you think it's a problem?

**Comment by lindenb** [comment]
My problem is solved  if I move the creation of the Anonymous class to a static method between `PARSER_BEGIN/PARSER_END` . Sadly, it's extra code but it works.

> Why do you think it's a problem?

@new-javacc javac isn't able  to compile this _(test must return a boolean)_:

`@Override public boolean test(final SAMRecord rec) { {if ("" != null) return rec.getDuplicateReadFlag();}}`

**Comment by new-javacc** [comment]
YEah - that's what I was wondering. If you look at your original 
grammar, you have a 4 otpions so the else part in the generated code 
should be throwing a parse exception. So that should work fine.

**Comment by new-javacc** [comment]
Just send me the generated code snippet for that method that causes 
error and I will what's going on.

**Comment by lindenb** [comment]
@new-javacc thanks, I'm away from my code for now, I'll send it tomorrow.
Otherwise, the jj file, before I moved the anonmous class to static, is here: 

https://github.com/lindenb/jvarkit/blob/a86c8971fe5ebb3f8de175c75e78f2d0e5325cfd/src/main/resources/javacc/com/github/lindenb/jvarkit/util/bio/samfilter/SamFilterParser.jj#L273

**Comment by theodore-norvell** [comment]
I still think this is an issue.  Why should the user need to use legacy exception handling to suppress the insertion of bad code?  The fact that the user needs to do a work-around indicates that there is a bug that needs to be worked around.

It seems to me that JavaCC is blindly rewriting all returns inside of Java blocks regardless of what subroutine the returns are returning from.  I'm sure the intention was to rewrite returns that return from the generated subroutine, not from subroutines that are nested inside of the generated one.

**Comment by zosrothko** [comment]
please ask @ainsley  his view on this issue. He made the changes on the exception handling.

**Comment by new-javacc** [comment]
Yeah remember the code generator is dumb - it just tokenizes the Java code so it can't distinguish those two cases. With my new code generator (when it's done :)) this will be moot. And like I said I strongly suggest keeping the action code really simple like function call and do all complex processing outside the grammar itself.But I do agree this is an issue with no quick fix.On Apr 19, 2017 8:18 AM, Theodore Norvell <notifications@github.com> wrote:I still think this is an issue.  Why should the user need to use legacy exception handling to suppress the insertion of bad code?  The fact that the user needs to do a work-around indicates that there is a bug that needs to be worked around.
It seems to me that JavaCC is blindly rewriting all returns inside of Java blocks regardless of what subroutine the returns are returning from.  I'm sure the intention was to rewrite returns that return from the generated subroutine, not from subroutines that are nested inside of the generated one.

**Comment by lindenb** [comment]
as requested, here is a minimal example:

```
options {
STATIC=false;
JDK_VERSION="1.8";
}

PARSER_BEGIN(Test)
import java.util.function.*;
public class Test {}
PARSER_END(Test)

TOKEN :{ <X: "x">}

Predicate<String> input():{}
	{
	<X> {
		return new Predicate<String>() {
			@Override public boolean test(final String s) { return s!=null;}
			};
		}
	}	
```

output:

```
$ java -cp javacc-release_7_0_2/bootstrap/javacc.jar javacc Test.jj  && javac Test.java 
Java Compiler Compiler Version 4.1d1 (Parser Generator)
(type "javacc" with no arguments for help)
Reading from file Test.jj . . .
File "TokenMgrError.java" is being rebuilt.
File "ParseException.java" is being rebuilt.
File "Token.java" is being rebuilt.
File "SimpleCharStream.java" is being rebuilt.
Parser generated successfully.
Test.java:8: error: missing return statement
                        @Override public boolean test(final String s) { {if (true) return s!=null;}}
                                                                                                   ^
Note: Test.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
1 error
```

**Comment by new-javacc** [comment]
Ok thanks! There might be a quick fix by changing it to also add: else throw errorTo every return in the code generator. Let me see if I can do it quickly.On Apr 19, 2017 10:01 AM, Pierre Lindenbaum <notifications@github.com> wrote:as requested, here is a minimal example:
options {
STATIC=false;
JDK_VERSION="1.8";
}

PARSER_BEGIN(Test)
import java.util.function.*;
public class Test {}
PARSER_END(Test)

TOKEN :{ <X: "x">}

Predicate<String> input():{}
	{
	<X> {
		return new Predicate<String>() {
			@Override public boolean test(final String s) { return s!=null;}
			};
		}
	}	

output:
$ java -cp javacc-release_7_0_2/bootstrap/javacc.jar javacc Test.jj  && javac Test.java 
Java Compiler Compiler Version 4.1d1 (Parser Generator)
(type "javacc" with no arguments for help)
Reading from file Test.jj . . .
File "TokenMgrError.java" is being rebuilt.
File "ParseException.java" is being rebuilt.
File "Token.java" is being rebuilt.
File "SimpleCharStream.java" is being rebuilt.
Parser generated successfully.
Test.java:8: error: missing return statement
                        @Override public boolean test(final String s) { {if (true) return s!=null;}}
                                                                                                   ^
Note: Test.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
1 error
